PWGHF/Tasks/taskMcEfficiency.cxx

```diff
@@ -0,0 +1,290 @@
+  void candidateLoop(T1& candidates, T2& tracks, T3& mcParticles)
+  {
+    using TracksType = std::decay_t<decltype(tracks)>;
+
+    auto hCandidates = registry.get<StepTHn>(HIST("hCandidates"));
+
+    for (auto pdgCode : (std::vector<int>)pdgCodes) {
+      auto decayType = -1;
+      std::array<int, 2> pdgDaughters;
+
+      if (pdgCode == pdg::kD0) {
+        decayType = 1 << DecayType::D0ToPiK;
+        pdgDaughters[0] = +kPiPlus;
+        pdgDaughters[1] = -kKPlus;
+      } else if (pdgCode == pdg::kD0Bar) {
+        decayType = 1 << DecayType::D0ToPiK;
+        pdgDaughters[0] = -kPiPlus;
+        pdgDaughters[1] = +kKPlus;
+      } else {
+        LOGP(fatal, "Not implemented for PDG {}", pdgCode);
+      }
+
+      std::map<int64_t, int> duplicates;
+
+      for (auto& candidate : selectedD0Candidates) {
+        if (!(candidate.hfflag() & decayType)) {
+          continue;
+        }
+
+        bool collisionMatched = false;
+        if constexpr (mc) {
+          auto trackPos = candidate.prong0_as<TracksType>();
+          auto trackNeg = candidate.prong1_as<TracksType>();
+
+          int8_t sign = 0;
+          auto indexRec = RecoDecay::getMatchedMCRec(mcParticles, std::array{trackPos, trackNeg}, pdgCode, pdgDaughters, false, &sign);
+          if (indexRec < 0) {
+            continue;
+          }
+
+          collisionMatched = candidate.collision_as<aod::McCollisionLabels>().mcCollisionId() == mcParticles.iteratorAt(indexRec).mcCollisionId();
+        }
+
+        float mass = -1;
+        bool selected = false;
+        if (pdgCode == pdg::kD0) {
+          mass = invMassD0ToPiK(candidate);
+          selected = candidate.isSelD0() >= selectionFlagD0;
+        } else if (pdgCode == pdg::kD0Bar) {
+          mass = invMassD0barToKPi(candidate);
+          selected = candidate.isSelD0bar() >= selectionFlagD0bar;
+        }
+        LOGP(debug, "Candidate {} has prong {} and prong {} and pT {} and mass {}", candidate.globalIndex(), candidate.prong0Id(), candidate.prong1Id(), candidate.pt(), mass);
+
+        // all candidates
+        hCandidates->Fill(kHFStepTracked, candidate.pt(), mass, pdgCode, candidate.cpa(), collisionMatched);
+
+        // check if prongs have passed track cuts
+        if (checkTrack(candidate.prong0_as<TracksType>()) && checkTrack(candidate.prong1_as<TracksType>())) {
+          hCandidates->Fill(kHFStepTrackedCuts, candidate.pt(), mass, pdgCode, candidate.cpa(), collisionMatched);
+        }
+
+        if (!selected) {
+          continue;
+        }
+
+        // selected candidates
+        hCandidates->Fill(kHFStepTrackedSelected, candidate.pt(), mass, pdgCode, candidate.cpa(), collisionMatched);
+
+        // duplicates
+        int64_t hash = 0;
+        if (candidate.prong0Id() < candidate.prong1Id()) {
+          hash = ((int64_t)candidate.prong0Id() << 32) | candidate.prong1Id();
+        } else {
+          hash = ((int64_t)candidate.prong1Id() << 32) | candidate.prong0Id();
```

Comment:
Would it help to move the `candidate.prong0Id()`, `candidate.prong1Id()` calls out and store the indices in variables?

Reply:
Not CPU wise. I kept this.